In [445]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
plt.style.use('seaborn-white')

In [446]:
df = pd.read_csv('customer_supermarket.csv', sep='\t')
df.drop(columns=['Unnamed: 0'], inplace=True)
df

BasketID      BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0        536365  01/12/10 08:26  2,55     17850.0  United Kingdom  85123A   
1        536365  01/12/10 08:26  3,39     17850.0  United Kingdom   71053   
2        536365  01/12/10 08:26  2,75     17850.0  United Kingdom  84406B   
3        536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029G   
4        536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029E   
...         ...             ...   ...         ...             ...     ...   
471905   581587  09/12/11 12:50   2,1     12680.0          France   22899   
471906   581587  09/12/11 12:50  4,15     12680.0          France   23254   
471907   581587  09/12/11 12:50  4,15     12680.0          France   23255   
471908   581587  09/12/11 12:50  4,95     12680.0          France   22138   
471909   581587  09/12/11 12:50    18     12680.0          France    POST   

                                  ProdDescr  Qta  
0        WHITE HANGING HEART T-LIGHT HOLDER    6  
1                       WHITE METAL LANTERN    6  
2            CREAM CUPID HEARTS COAT HANGER    8  
3       KNITTED UNION FLAG HOT WATER BOTTLE    6  
4            RED WOOLLY HOTTIE WHITE HEART.    6  
...                                     ...  ...  
471905         CHILDREN'S APRON DOLLY GIRL     6  
471906        CHILDRENS CUTLERY DOLLY GIRL     4  
471907      CHILDRENS CUTLERY CIRCUS PARADE    4  
471908        BAKING SET 9 PIECE RETROSPOT     3  
471909                              POSTAGE    1  

[471910 rows x 8 columns]

In [447]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 471910 entries, 0 to 471909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         471910 non-null  object 
 1   BasketDate       471910 non-null  object 
 2   Sale             471910 non-null  object 
 3   CustomerID       406830 non-null  float64
 4   CustomerCountry  471910 non-null  object 
 5   ProdID           471910 non-null  object 
 6   ProdDescr        471157 non-null  object 
 7   Qta              471910 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 28.8+ MB


BasketID è una stringa, contiene lettere C che indicano la cancellazione. Eliminiamo le righe corrispondenti. 

In [448]:
import re

def search_letters(the_string):
    if re.search('[a-zA-Z]', the_string) is None:
        return False
    return True

df[df['BasketID'].apply(search_letters)]

BasketID      BasketDate    Sale  CustomerID CustomerCountry  ProdID  \
141     C536379  01/12/10 09:41    27,5     14527.0  United Kingdom       D   
154     C536383  01/12/10 09:49    4,65     15311.0  United Kingdom  35004C   
235     C536391  01/12/10 10:24    1,65     17548.0  United Kingdom   22556   
236     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21984   
237     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21983   
...         ...             ...     ...         ...             ...     ...   
470979  C581490  09/12/11 09:57    0,83     14397.0  United Kingdom   23144   
471541  C581499  09/12/11 10:28  224,69     15498.0  United Kingdom       M   
471715  C581568  09/12/11 11:57   10,95     15311.0  United Kingdom   21258   
471716  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   84978   
471717  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   20979   

                               ProdDescr  Qta  
141                             Discount   -1  
154      SET OF 3 COLOURED  FLYING DUCKS   -1  
235       PLASTERS IN TIN CIRCUS PARADE   -12  
236     PACK OF 12 PINK PAISLEY TISSUES   -24  
237     PACK OF 12 BLUE PAISLEY TISSUES   -24  
...                                  ...  ...  
470979   ZINC T-LIGHT HOLDER STARS SMALL  -11  
471541                            Manual   -1  
471715        VICTORIAN SEWING BOX LARGE   -5  
471716  HANGING HEART JAR T-LIGHT HOLDER   -1  
471717     36 PENCILS TUBE RED RETROSPOT   -5  

[9086 rows x 8 columns]

In [449]:
def search_letter_C(the_string):
    if re.search('[C]', the_string) is None:
        return False
    return True

df[df['BasketID'].apply(search_letter_C)]

BasketID      BasketDate    Sale  CustomerID CustomerCountry  ProdID  \
141     C536379  01/12/10 09:41    27,5     14527.0  United Kingdom       D   
154     C536383  01/12/10 09:49    4,65     15311.0  United Kingdom  35004C   
235     C536391  01/12/10 10:24    1,65     17548.0  United Kingdom   22556   
236     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21984   
237     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21983   
...         ...             ...     ...         ...             ...     ...   
470979  C581490  09/12/11 09:57    0,83     14397.0  United Kingdom   23144   
471541  C581499  09/12/11 10:28  224,69     15498.0  United Kingdom       M   
471715  C581568  09/12/11 11:57   10,95     15311.0  United Kingdom   21258   
471716  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   84978   
471717  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   20979   

                               ProdDescr  Qta  
141                             Discount   -1  
154      SET OF 3 COLOURED  FLYING DUCKS   -1  
235       PLASTERS IN TIN CIRCUS PARADE   -12  
236     PACK OF 12 PINK PAISLEY TISSUES   -24  
237     PACK OF 12 BLUE PAISLEY TISSUES   -24  
...                                  ...  ...  
470979   ZINC T-LIGHT HOLDER STARS SMALL  -11  
471541                            Manual   -1  
471715        VICTORIAN SEWING BOX LARGE   -5  
471716  HANGING HEART JAR T-LIGHT HOLDER   -1  
471717     36 PENCILS TUBE RED RETROSPOT   -5  

[9084 rows x 8 columns]

In [450]:
def search_letters_noC(the_string):
    if re.search('[a-zA-Z]', the_string) is None:
        return False
    elif re.search('[C]', the_string) is None:
        return True
    return False

df[df['BasketID'].apply(search_letters_noC)]

BasketID      BasketDate       Sale  CustomerID CustomerCountry ProdID  \
256523  A563186  12/08/11 14:51  -11062,06         NaN  United Kingdom      B   
256524  A563187  12/08/11 14:52  -11062,06         NaN  United Kingdom      B   

              ProdDescr  Qta  
256523  Adjust bad debt    1  
256524  Adjust bad debt    1

In [451]:
tmp = df[df['BasketID'].apply(lambda s: not search_letters(s))]
df = tmp.copy()

Abbiamo eliminato 9086 righe (tutte quelle che hanno la lettera C o A davanti a BasketID).

In [452]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 462824 entries, 0 to 471909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         462824 non-null  object 
 1   BasketDate       462824 non-null  object 
 2   Sale             462824 non-null  object 
 3   CustomerID       397925 non-null  float64
 4   CustomerCountry  462824 non-null  object 
 5   ProdID           462824 non-null  object 
 6   ProdDescr        462071 non-null  object 
 7   Qta              462824 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 31.8+ MB


Trasformo BasketID in intero.

In [453]:
df['BasketID'] = df['BasketID'].astype(int).copy()

In [454]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 462824 entries, 0 to 471909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         462824 non-null  int64  
 1   BasketDate       462824 non-null  object 
 2   Sale             462824 non-null  object 
 3   CustomerID       397925 non-null  float64
 4   CustomerCountry  462824 non-null  object 
 5   ProdID           462824 non-null  object 
 6   ProdDescr        462071 non-null  object 
 7   Qta              462824 non-null  int64  
dtypes: float64(1), int64(2), object(5)
memory usage: 31.8+ MB


In [455]:
df[df['Qta'] <= 0]

BasketID      BasketDate Sale  CustomerID CustomerCountry  ProdID  \
3738      536764  02/12/10 14:42    0         NaN  United Kingdom  84952C   
6029      536996  03/12/10 15:30    0         NaN  United Kingdom   22712   
6030      537000  03/12/10 15:32    0         NaN  United Kingdom   21414   
6031      537001  03/12/10 15:33    0         NaN  United Kingdom   21653   
6032      537003  03/12/10 15:33    0         NaN  United Kingdom   85126   
...          ...             ...  ...         ...             ...     ...   
467098    581198  07/12/11 18:26    0         NaN  United Kingdom   22025   
467099    581199  07/12/11 18:26    0         NaN  United Kingdom   84581   
467100    581201  07/12/11 18:30    0         NaN  United Kingdom   22217   
467103    581205  07/12/11 18:32    0         NaN  United Kingdom   20893   
467106    581212  07/12/11 18:38    0         NaN  United Kingdom   22578   

       ProdDescr   Qta  
3738         NaN   -38  
6029         NaN   -20  
6030         NaN   -22  
6031         NaN    -6  
6032         NaN    -2  
...          ...   ...  
467098     check   -30  
467099       NaN    -2  
467100  damages?  -155  
467103   damages   -55  
467106      lost -1050  

[668 rows x 8 columns]

In [456]:
df[df['Qta'] < 0]['Sale'].value_counts()

0    668
Name: Sale, dtype: int64

Tutti i record con Qta negativa hanno Sale = 0, allora gli eliminiamo in quanto non sono interessati per l'analisi. Inoltre sono solamente 668 righe.

In [457]:
df = df[df['Qta'] > 0].copy()

In [458]:
df.shape

(462156, 8)

In [459]:
def convert_to_float(s):
    comma_pos = s.find(',')
    if comma_pos >= 0:
        s = s[: comma_pos] + '.' + s[comma_pos+1 :]
    return float(s)

df['Sale'] = df['Sale'].apply(convert_to_float).copy()

In [460]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 462156 entries, 0 to 471909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         462156 non-null  int64  
 1   BasketDate       462156 non-null  object 
 2   Sale             462156 non-null  float64
 3   CustomerID       397925 non-null  float64
 4   CustomerCountry  462156 non-null  object 
 5   ProdID           462156 non-null  object 
 6   ProdDescr        461850 non-null  object 
 7   Qta              462156 non-null  int64  
dtypes: float64(2), int64(2), object(4)
memory usage: 31.7+ MB


In [461]:
df = df[df['Sale'] > 0].copy()
df.shape

(461545, 8)

In [462]:
df[df['ProdID'].apply(str.isalpha)]

BasketID      BasketDate     Sale  CustomerID CustomerCountry ProdID  \
45        536370  01/12/10 08:45    18.00     12583.0          France   POST   
386       536403  01/12/10 11:27    15.00     12791.0     Netherlands   POST   
1122      536527  01/12/10 13:04    18.00     12662.0         Germany   POST   
1976      536569  01/12/10 15:35     1.25     16274.0  United Kingdom      M   
1987      536569  01/12/10 15:35    18.95     16274.0  United Kingdom      M   
...          ...             ...      ...         ...             ...    ...   
471540    581498  09/12/11 10:26  1714.17         NaN  United Kingdom    DOT   
471730    581570  09/12/11 11:59    18.00     12662.0         Germany   POST   
471767    581574  09/12/11 12:09    18.00     12526.0         Germany   POST   
471768    581578  09/12/11 12:16    18.00     12713.0         Germany   POST   
471909    581587  09/12/11 12:50    18.00     12680.0          France   POST   

             ProdDescr  Qta  
45             POSTAGE    3  
386            POSTAGE    1  
1122           POSTAGE    1  
1976            Manual    1  
1987            Manual    1  
...                ...  ...  
471540  DOTCOM POSTAGE    1  
471730         POSTAGE    1  
471767         POSTAGE    2  
471768         POSTAGE    3  
471909         POSTAGE    1  

[1758 rows x 8 columns]

1758 record hanno 'ProdID' con solo lettere e non rappresentano veri e propri acquisti.

In [463]:
df[df['ProdID'].apply(str.isalpha)]['ProdID'].value_counts()

POST         1111
DOT           335
M             296
DCGSSBOY        6
DCGSSGIRL       4
PADS            3
AMAZONFEE       1
S               1
m               1
Name: ProdID, dtype: int64

In [464]:
df = df[df['ProdID'].apply(lambda s: not s.isalpha())].copy()

In [465]:
def search_letters(the_string):
    if re.search('[a-z]', the_string) is None:
        return False
    return True

strange_prodID = df['ProdID'].apply(search_letters)
df[df['ProdID'].apply(search_letters)]['ProdID'].value_counts().index.values

array(['47566b', '15056n', '72349b', '84509a', '84997c', '82494l',
       '84997b', '84997a', '84997d', '85123a', '84970s', '15056bl',
       '84970l', '84030e', '85049g', '47591d', '85049a', '72801d',
       '84510a', '85049e', '85039a', '84884a', '15056p', '85040a',
       '85014b', '85114c', '84993a', '84872a', '47559b', '85114b',
       '84559b', '84559a', '84989a', '84509c', '84660c', '84510c',
       '72801c', '84796a', '18098c', '85132b', '85114a', '82001s',
       '85014a', '72351a', '84596f', '85099f', 'gift_0001_30', '84968e',
       '47518f', '47590b', '85132a', '84596b', 'gift_0001_40', '85231g',
       '84968a', '84596e', '79066k', 'gift_0001_10', '72807c', '35809a',
       'gift_0001_20', '84536a', 'gift_0001_50', '85049d', '72807b',
       '48173c', '84795b', '85231b', '84509b', '47570b', '46776e',
       '72807a', '84968d', '84661b', '85035b', '84971l', '85035a',
       '84536b', '51014c', '85184c', '84660b', '84031a', '90184c',
       '82613b', '84031b', '84661a', '850

In [466]:
df = df[df['ProdID'].apply(lambda s: s.find('gift_0001_') == -1)].copy()
df

BasketID      BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0         536365  01/12/10 08:26  2.55     17850.0  United Kingdom  85123A   
1         536365  01/12/10 08:26  3.39     17850.0  United Kingdom   71053   
2         536365  01/12/10 08:26  2.75     17850.0  United Kingdom  84406B   
3         536365  01/12/10 08:26  3.39     17850.0  United Kingdom  84029G   
4         536365  01/12/10 08:26  3.39     17850.0  United Kingdom  84029E   
...          ...             ...   ...         ...             ...     ...   
471904    581587  09/12/11 12:50  0.85     12680.0          France   22613   
471905    581587  09/12/11 12:50  2.10     12680.0          France   22899   
471906    581587  09/12/11 12:50  4.15     12680.0          France   23254   
471907    581587  09/12/11 12:50  4.15     12680.0          France   23255   
471908    581587  09/12/11 12:50  4.95     12680.0          France   22138   

                                  ProdDescr  Qta  
0        WHITE HANGING HEART T-LIGHT HOLDER    6  
1                       WHITE METAL LANTERN    6  
2            CREAM CUPID HEARTS COAT HANGER    8  
3       KNITTED UNION FLAG HOT WATER BOTTLE    6  
4            RED WOOLLY HOTTIE WHITE HEART.    6  
...                                     ...  ...  
471904          PACK OF 20 SPACEBOY NAPKINS   12  
471905         CHILDREN'S APRON DOLLY GIRL     6  
471906        CHILDRENS CUTLERY DOLLY GIRL     4  
471907      CHILDRENS CUTLERY CIRCUS PARADE    4  
471908        BAKING SET 9 PIECE RETROSPOT     3  

[459772 rows x 8 columns]

In [467]:
df['ProdDescr'].value_counts()

WHITE HANGING HEART T-LIGHT HOLDER    2169
JUMBO BAG RED RETROSPOT               1871
REGENCY CAKESTAND 3 TIER              1859
PARTY BUNTING                         1548
ASSORTED COLOUR BIRD ORNAMENT         1459
                                      ... 
SET 10 CARDS MAGICAL TREE 17086          1
HEAVENS SCENT FRAGRANCE OILS ASSTD       1
PINK PAINTED KASHMIRI CHAIR              1
WHITE WITH METAL BAG CHARM               1
MARIE ANTOIENETT TRINKET BOX GOLD        1
Name: ProdDescr, Length: 3977, dtype: int64

In [468]:
df[df['ProdID'] == 'C2']

BasketID      BasketDate  Sale  CustomerID  CustomerCountry ProdID  \
1422      536540  01/12/10 14:05  50.0     14911.0             EIRE     C2   
10319     537368  06/12/10 12:40  50.0     14911.0             EIRE     C2   
10652     537378  06/12/10 13:06  50.0     14911.0             EIRE     C2   
16525     537963  09/12/10 11:30  50.0     13369.0   United Kingdom     C2   
16564     538002  09/12/10 11:48  50.0     14932.0  Channel Islands     C2   
...          ...             ...   ...         ...              ...    ...   
447963    579539  30/11/11 10:06  50.0     14911.0             EIRE     C2   
449563    579768  30/11/11 15:08  50.0     14911.0             EIRE     C2   
450596    579910  01/12/11 08:52  50.0     14911.0             EIRE     C2   
452666    580127  01/12/11 17:51  50.0     14911.0             EIRE     C2   
457721    580555  05/12/11 10:18  50.0     14911.0             EIRE     C2   

       ProdDescr  Qta  
1422    CARRIAGE    1  
10319   CARRIAGE    1  
10652   CARRIAGE    1  
16525   CARRIAGE    1  
16564   CARRIAGE    1  
...          ...  ...  
447963  CARRIAGE    1  
449563  CARRIAGE    1  
450596  CARRIAGE    1  
452666  CARRIAGE    1  
457721  CARRIAGE    1  

[139 rows x 8 columns]

In [469]:
df = df[df['ProdID'] != 'C2'].copy()
df.shape

(459633, 8)

In [470]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 459633 entries, 0 to 471908
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         459633 non-null  int64  
 1   BasketDate       459633 non-null  object 
 2   Sale             459633 non-null  float64
 3   CustomerID       396349 non-null  float64
 4   CustomerCountry  459633 non-null  object 
 5   ProdID           459633 non-null  object 
 6   ProdDescr        459633 non-null  object 
 7   Qta              459633 non-null  int64  
dtypes: float64(2), int64(2), object(4)
memory usage: 31.6+ MB


In [471]:
from dateutil.parser import parse

def is_date(string, fuzzy=False):
    """
    Return whether the string can be interpreted as a date.

    :param string: str, string to check for date
    :param fuzzy: bool, ignore unknown tokens in string if True
    """
    try: 
        parse(string, fuzzy=fuzzy)
        return True

    except ValueError:
        return False


is_date_output = df['BasketDate'].apply(is_date)
is_date_output

0         True
1         True
2         True
3         True
4         True
          ... 
471904    True
471905    True
471906    True
471907    True
471908    True
Name: BasketDate, Length: 459633, dtype: bool

In [472]:
is_date_output.value_counts()

True    459633
Name: BasketDate, dtype: int64

All dates are well formatted.

In [473]:
import datetime

def validate(date_text):
    try:
        datetime.datetime.strptime(date_text, '%d/%m/%y %H:%M')
        return True
    except ValueError:
        return False

    
correct_date_format = df['BasketDate'].apply(validate)
correct_date_format.value_counts()

True    459633
Name: BasketDate, dtype: int64

All dates are in the same format.

In [474]:
df['BasketDate'] = pd.to_datetime(df['BasketDate'], format="%d/%m/%y %H:%M").copy()

In [475]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 459633 entries, 0 to 471908
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   BasketID         459633 non-null  int64         
 1   BasketDate       459633 non-null  datetime64[ns]
 2   Sale             459633 non-null  float64       
 3   CustomerID       396349 non-null  float64       
 4   CustomerCountry  459633 non-null  object        
 5   ProdID           459633 non-null  object        
 6   ProdDescr        459633 non-null  object        
 7   Qta              459633 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(2), object(3)
memory usage: 31.6+ MB


In [476]:
df[df['CustomerCountry'] == 'Unspecified']

BasketID          BasketDate  Sale  CustomerID CustomerCountry ProdID  \
128523    549687 2011-04-11 13:29:00  7.95     12363.0     Unspecified  20685   
128524    549687 2011-04-11 13:29:00  7.95     12363.0     Unspecified  22691   
128525    549687 2011-04-11 13:29:00  7.95     12363.0     Unspecified  48116   
128526    549687 2011-04-11 13:29:00  0.55     12363.0     Unspecified  21213   
128527    549687 2011-04-11 13:29:00  0.55     12363.0     Unspecified  21977   
...          ...                 ...   ...         ...             ...    ...   
435421    578539 2011-11-24 14:55:00  4.95         NaN     Unspecified  23569   
435422    578539 2011-11-24 14:55:00  0.42         NaN     Unspecified  23507   
435423    578539 2011-11-24 14:55:00  1.25         NaN     Unspecified  23570   
435424    578539 2011-11-24 14:55:00  1.65         NaN     Unspecified  23571   
435425    578539 2011-11-24 14:55:00  0.55         NaN     Unspecified  23309   

                                  ProdDescr  Qta  
128523                DOORMAT RED RETROSPOT    2  
128524              DOORMAT WELCOME SUNRISE    2  
128525           DOORMAT MULTICOLOUR STRIPE    2  
128526          PACK OF 72 SKULL CAKE CASES   24  
128527   PACK OF 60 PINK PAISLEY CAKE CASES   24  
...                                     ...  ...  
435421        TRADTIONAL ALPHABET STAMP SET    4  
435422     MINI PLAYING CARDS BUFFALO BILL    20  
435423     TRADITIONAL PICK UP STICKS GAME    12  
435424        TRADITIONAL NAUGHTS & CROSSES   12  
435425  SET OF 60 I LOVE LONDON CAKE CASES    24  

[340 rows x 8 columns]

In [487]:
null = df[df['CustomerID'].isnull()]['BasketID'].value_counts()
null

573585    534
580729    357
581219    351
581492    351
558475    330
         ... 
570286      1
566761      1
541118      1
551847      1
557610      1
Name: BasketID, Length: 1253, dtype: int64

In [488]:
non_null = df.dropna()['BasketID'].value_counts()
non_null

576339    541
579196    532
580727    528
578270    441
573576    434
         ... 
545448      1
571237      1
577848      1
572085      1
556249      1
Name: BasketID, Length: 18413, dtype: int64

In [489]:
null.index.values

array([573585, 580729, 581219, ..., 541118, 551847, 557610])

In [490]:
non_null.index.values

array([576339, 579196, 580727, ..., 577848, 572085, 556249])

In [492]:
def intersection(lst1, lst2): 
    return list(set(lst1) & set(lst2))

intersection(list(null.index.values), list(non_null.index.values))

[]

Non si possono recuperare i CustomerID dai BasketID.

In [493]:
df['CustomerID'].value_counts()

17841.0    7838
14911.0    5589
14096.0    5095
12748.0    4580
14606.0    2698
           ... 
13302.0       1
16738.0       1
13106.0       1
16257.0       1
14424.0       1
Name: CustomerID, Length: 4335, dtype: int64

In [495]:
df[df['CustomerID'].isnull()]['CustomerCountry'].value_counts()

United Kingdom    62634
EIRE                297
Hong Kong           134
Unspecified          96
Switzerland          55
France               33
Israel               19
Portugal             15
Bahrain               1
Name: CustomerCountry, dtype: int64